src/client/dfs/duns.c

```diff
@@ -0,0 +1,236 @@
+		else if (s > DUNS_MAX_XATTR_LEN)
+			D_ERROR("Invalid xattr length\n");
+		else
+			D_ERROR("Invalid DAOS unified namespace xattr\n");
+		return -DER_INVAL;
+	}
+
+	t = strtok_r(str, ".", &saveptr);
+	if (t == NULL) {
+		fprintf(stderr, "Invalid DAOS xattr format (%s).\n", str);
+		return -DER_INVAL;
+	}
+
+	t = strtok_r(NULL, ":", &saveptr);
+	daos_parse_ctype(t, &attr->da_type);
+	if (attr->da_type == DAOS_PROP_CO_LAYOUT_UNKOWN) {
+		D_ERROR("Invalid DAOS xattr format: Container layout cannot be"
+			" unknown\n");
+		return -DER_INVAL;
+	}
+
+	t = strtok_r(NULL, "/", &saveptr);
+	rc = uuid_parse(t, attr->da_puuid);
+	if (rc) {
+		D_ERROR("Invalid DAOS xattr format: pool UUID cannot be"
+			" parsed\n");
+		return -DER_INVAL;
+	}
+
+	t = strtok_r(NULL, "/", &saveptr);
+	rc = uuid_parse(t, attr->da_cuuid);
+	if (rc) {
+		D_ERROR("Invalid DAOS xattr format: container UUID cannot be"
+			" parsed\n");
+		return -DER_INVAL;
+	}
+
+	t = strtok_r(NULL, "/", &saveptr);
+	daos_parse_oclass(t, &attr->da_oclass);
+
+	return 0;
+}
+
+int
+duns_link_path(const char *path, struct duns_attr_t attr)
+{
+	char			*svcl_str, *group;
+	d_rank_list_t		*svcl = NULL;
+	daos_handle_t		poh;
+	daos_pool_info_t	pool_info;
+	char			pool[37], cont[37];
+	char			oclass[10], type[10];
+	char			str[DUNS_MAX_XATTR_LEN];
+	int			len;
+	int			rc;
+
+	if (path == NULL) {
+		D_ERROR("Invalid path\n");
+		return -DER_INVAL;
+	}
+
+	/*
+	 * MSC - for now SVCL and group need to be passed through env
+	 * variable. they shouldn't be required later.
+	 */
+	svcl_str = getenv("DAOS_SVCL");
+	if (svcl_str == NULL) {
+		fprintf(stderr, "missing pool service rank list\n");
+		return -DER_INVAL;
+	}
+
+	svcl = daos_rank_list_parse(svcl_str, ":");
+	if (svcl == NULL) {
+		fprintf(stderr, "Invalid pool service rank list\n");
+		return -DER_INVAL;
+	}
+
+	group = getenv("DAOS_GROUP");
+
+	if (attr.da_type == DAOS_PROP_CO_LAYOUT_HDF5) {
+		/** create a new file if HDF5 container */
+		int fd;
+
+		fd = open(path, O_CREAT | O_EXCL,
+			  S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP | S_IROTH);
+		if (fd == -1) {
+			D_ERROR("Failed to create file %s: %s\n", path,
+				strerror(errno));
+			/** TODO - convert errno to rc */
+			return -DER_INVAL;
+		}
+		close(fd);
+	} else if (attr.da_type == DAOS_PROP_CO_LAYOUT_POSIX) {
+		/** create a new directory if POSIX/MPI-IO container */
+		rc = mkdir(path, S_IRWXU | S_IRWXG | S_IROTH | S_IWOTH);
+		if (rc == -1) {
+			D_ERROR("Failed to create dir %s: %s\n",
+				path, strerror(errno));
+			/** TODO - convert errno to rc */
+			return -DER_INVAL;
+		}
+	} else {
+		D_ERROR("Invalid container layout.\n");
+		return -DER_INVAL;
+	}
+
+	/** Connect to the pool. */
+	rc = daos_pool_connect(attr.da_puuid, group, svcl, DAOS_PC_RW, &poh,
+			       &pool_info, NULL);
+	if (rc) {
+		D_ERROR("Failed to connect to pool (%d)\n", rc);
+		D_GOTO(err_link, rc);
+	}
+
+	/** generate a random container uuid and try to create it */
+	do {
+		uuid_generate(attr.da_cuuid);
```

Comment:
for recovery purpose, i would set the xattr (and sync it) on the file before calling daos_cont_create(). This will make sure that we are never leaking containers. Anyway, can be a follow-up improvements.

Reply:
done